File: sol/871.cpp
#include <queue>

struct Solution {
    int minRefuelStops(int target, int startFuel, vector<vector<int>>& stations) {
        int n = stations.size();
        if (startFuel >= target){
            return 0;
        }
        priority_queue<int> pq;
        long maxReach = startFuel;
        int maxIdx = 0;
        for (int k = 1; k <= n; ++k){
            while (maxIdx < n && stations[maxIdx][0] <= maxReach){
                pq.push(stations[maxIdx++][1]);
            }
            if (pq.empty()){
                return -1;
            }
            maxReach += pq.top();
            pq.pop();
            if (maxReach >= target){
                return k;
            }
        }
        return -1;
    }
};

File: sol/871/871.py
import heapq as hq

class Solution:
    def minRefuelStops(self, target: int, startFuel: int, stations: List[List[int]]) -> int:
        n = len(stations)
        if startFuel >= target:
            return 0
        h = []
        maxReach = startFuel
        maxIdx = 0
        for k in range(1, n + 1):
            while maxIdx < n and stations[maxIdx][0] <= maxReach:
                hq.heappush(h, -stations[maxIdx][1])
                maxIdx += 1
            if len(h) == 0:
                return -1
            maxReach -= hq.heappop(h)
            if maxReach >= target:
                return k
        return -1

